packages/profile-sync-controller/src/UserStorageController.ts

```diff
@@ -0,0 +1,397 @@
+import type {
+  RestrictedControllerMessenger,
+  StateMetadata,
+} from '@metamask/base-controller';
+import { BaseController } from '@metamask/base-controller';
+import type {
+  MetamaskNotificationsControllerDisableMetamaskNotifications,
+  MetamaskNotificationsControllerSelectIsMetamaskNotificationsEnabled,
+} from '@metamask/notifications-controller';
+import type { HandleSnapRequest } from '@metamask/snaps-controllers';
+
+import type {
+  AuthenticationControllerGetBearerToken,
+  AuthenticationControllerGetSessionProfile,
+  AuthenticationControllerIsSignedIn,
+  AuthenticationControllerPerformSignIn,
+  AuthenticationControllerPerformSignOut,
+} from '.';
+import { createSnapSignMessageRequest } from './AuthSnapRequests';
+import { createSHA256Hash } from './encryption';
+import type { UserStorageEntryKeys } from './schema';
+import {
+  getUserStorage,
+  upsertUserStorage,
+} from './services/user-storage-controller';
+
+const controllerName = 'UserStorageController';
+
+type SessionProfile = {
+  identifierId: string;
+  profileId: string;
+};
+
+// State
+export type UserStorageControllerState = {
+  /**
+   * Condition used by UI and to determine if we can use some of the User Storage methods.
+   */
+  isProfileSyncingEnabled: boolean;
+  /**
+   * Loading state for the profile syncing update
+   */
+  isProfileSyncingUpdateLoading: boolean;
+};
+
+const defaultState: UserStorageControllerState = {
+  isProfileSyncingEnabled: true,
+  isProfileSyncingUpdateLoading: false,
+};
+
+const metadata: StateMetadata<UserStorageControllerState> = {
+  isProfileSyncingEnabled: {
+    persist: true,
+    anonymous: true,
+  },
+  isProfileSyncingUpdateLoading: {
+    persist: false,
+    anonymous: false,
+  },
+};
+
+// Messenger Actions
+type CreateActionsObj<T extends keyof UserStorageController> = {
+  [K in T]: {
+    type: `${typeof controllerName}:${K}`;
+    handler: UserStorageController[K];
+  };
+};
+type ActionsObj = CreateActionsObj<
+  | 'performGetStorage'
+  | 'performSetStorage'
+  | 'getStorageKey'
+  | 'enableProfileSyncing'
+  | 'disableProfileSyncing'
+>;
+export type Actions = ActionsObj[keyof ActionsObj];
+export type UserStorageControllerPerformGetStorage =
```

Comment:
Similar as above — action types should end in "Action".

Reply:
addressed